src/irmin/tree.ml

```diff
@@ -1258,76 +1434,199 @@ module Make (P : Private.S) = struct
+        StepMap.to_seq x
+        |> Seq.map (fun (step, v) ->
+               match v with
+               | `Node n -> (
+                   match Node.cached_key n with
+                   | Some k -> (step, `Node k)
+                   | None ->
+                       assertion_failure
+                         "Encountered child node value with uncached key \
+                          during export:@,\
+                          @ @[%a@]" dump v)
+               | `Contents (c, m) -> (
+                   match Contents.cached_key c with
+                   | Some k -> (step, `Contents (k, m))
+                   | None ->
+                       assertion_failure
+                         "Encountered child contents value with uncached key \
+                          during export:@,\
+                          @ @[%a@]" dump v))
+        |> P.Node.Val.of_seq
+      in
+      add_node n node k
+    in
+
+    let add_updated_node n (v : Node.value) (updates : Node.updatemap) k =
+      let node =
+        StepMap.fold
+          (fun k v acc ->
+            match v with
+            | Node.Remove -> P.Node.Val.remove acc k
+            | Node.Add (`Node n as v) -> (
+                match Node.cached_key n with
+                | Some ptr -> P.Node.Val.add acc k (`Node ptr)
+                | None ->
+                    assertion_failure
+                      "Encountered child node value with uncached key during \
+                       export:@,\
+                       @ @[%a@]" dump v)
+            | Add (`Contents (c, m) as v) -> (
+                match Contents.cached_key c with
+                | Some ptr -> P.Node.Val.add acc k (`Contents (ptr, m))
+                | None ->
+                    assertion_failure
+                      "Encountered child contents value with uncached key \
+                       during export:@,\
+                       @ @[%a@]" dump v))
+          updates v
+      in
+      add_node n node k
     in
-    let rec on_node (`Node n) k =
+
+    let rec on_node : type r. [ `Node of node ] -> (unit, r) cont_lwt =
+     fun (`Node n) k ->
+      (* Invariant: when [k] is called, [Node.cached_key n = Some _]. *)
+      let k () =
+        (match Node.cached_key n with
+        | None ->
+            Log.err (fun f ->
+                f "After trying to export %a, key isn't cached." dump (`Node n));
+            assert false
+        | Some _ -> ());
+        k ()
+      in
       match n.Node.v with
-      | Node.Hash (_, h) ->
-          Node.export ?clear repo n h;
-          k ()
-      | Node.Value (_, v, None) ->
-          let h = P.Node.Key.hash v in
-          Node.export ?clear repo n h;
+      | Node.Key (_, key) ->
+          Node.export ?clear repo n key;
           k ()
       | _ -> (
           skip n >>= function
-          | true -> k ()
-          | false ->
+          | true ->
+              Log.err (fun f -> f "Skipping instead");
```

Comment:
I'm not sure I understand, is it an error to skip a node here?

Reply:
Nope, this is just a debug message that snuck in... :sweat_smile: